### PR TITLE
Implement makeTempDirSync()

### PR DESCRIPTION
**Pull request by piscisaureus**

### CODE REVIEW

**Comment by ry** [comment]
Here is prior art:

Node (tmp module)
```javascript
var tmp = require('tmp');
var tmpobj = tmp.dirSync();
console.log('Dir: ', tmpobj.name);
tmpobj.removeCallback(); // Manual cleanup
```
Go
```go
func TempDir(dir, prefix string) (name string, err error)
```
Rust
```rust
    let tmp_dir = TempDir::new("example")?;
    let file_path = tmp_dir.path().join("my-temporary-note.txt");
```
C
```c
#include <stdlib.h>
/* The mkdtemp() function generates a uniquely named temporary directory 
from template. The last six characters of template must be XXXXXX and these 
are replaced with a string that makes the directory name unique. The directory 
is then created with permissions 0700. Since it will be modified, template must 
not be a string constant, but should be declared as a character array. */
char *mkdtemp(char *template);
```
Python
```
tempfile.mkdtemp([suffix=''[, prefix='tmp'[, dir=None]]])
```
Ruby
```ruby
dir = Dir.mktmpdir
begin
  # use the directory...
  open("#{dir}/foo", "w") { ... }
ensure
  # remove the directory.
  FileUtils.remove_entry dir
end
```

**Comment by benjamingr** [comment]
> Node (tmp module)

```js
fs.makedtempSync(`{os.tmpdir()}{path.sep}`)
```

Is probably more idiomatic "modern" node I guess and similar to the C API.

**Comment by piscisaureus** [comment]
@ry, @benjamingr 

Context matters here. I think if we had a module "fs" or an "fs" object, it'd make sense to call this method "tempDir()". I prefer short names over long ones too. But at the moment makeTempDir() has to hang off the deno object, and internally it's in the OS module, which can make it much more ambiguous.

In Rust, `std::env::temp_dir()` returns the default temp dir, it doesn't create one.
In Go, `os.TempDir()`, same story.
In node, `os.tmpdir()` -> same.
In ruby, `Dir.tmpdir()` -> ditto.

Let's stick to this long and hopelessly boring name and revisit once we have some internal namespaces set up.

**Comment by ry** [comment]
@piscisaureus Sounds good. So we will use `deno.tempDir()` to get `$TMPDIR` ?

 I think Go's usage of os.tempDir vs ioutil.tempDir() is quite confusing and unnecessary. I want to have a flat namespace and just choose unique names for every op.

**Comment by ry** [comment]
We should be consistent with our usage of "make" then. 
Imaginary example elsewhere it might be suggested to add "createSocket" - we should not do that and instead do "makeSocket".  However, I think "mkdir" should not be "makeDir" but instead use the more common name.
Ideally we could codify these naming rules... @benjamingr, I'm hoping you could help us with making that document (ref #566)

**Comment by piscisaureus** [comment]
> We should be consistent with our usage of "make" then.

I agree that we should be *if* we stick to it. There's something to say for 'createTempDir' too, although it's even longer.
But it's hard to judge in isolation, maybe we can make/create a table of fs operations that we need to implement, so we can see the whole picture.

File: js/unit_tests.ts [hunk 1]
```diff
@@ -70,23 +70,45 @@ testPerm({ write: true }, function writeFileSyncFail() {
...
+test(function makeTempDirSyncPerm() {
+  // makeTempDirSync should require write permissions (for now).
+  let err;
+  try {
+    deno.makeTempDirSync({ dir: "/baddir" });
+  } catch (err_) {
+    err = err_;
+  }
+  console.log(err);
+  assert(err && err.name === "deno.PermissionDenied");
```
Comment (ry) [inline js/unit_tests.ts]: Add
```
// TODO assert(err instanceof deno.PermissionDenied)
```

File: js/unit_tests.ts [hunk 2]
```diff
@@ -70,23 +70,45 @@ testPerm({ write: true }, function writeFileSyncFail() {
...
+  assert(dir3.startsWith(dir1));
+  assert(/^[\\\/]/.test(dir3.slice(dir1.length)));
+  // Check that creating a temp dir inside a nonexisting directory fails.
+  let err;
+  try {
+    deno.makeTempDirSync({ dir: "/baddir" });
+  } catch (err_) {
+    err = err_;
+  }
+  assert(err && err.name === "deno.NotFound");
```
Comment (ry) [inline js/unit_tests.ts]:
```
// TODO assert(err instanceof deno.NotFound)
```

File: js/unit_tests.ts [hunk 3]
```diff
@@ -70,23 +70,45 @@ testPerm({ write: true }, function writeFileSyncFail() {
   assert(caughtError);
 });
 
+testPerm({ write: true }, function makeTempDirSync() {
```
Comment (ry) [inline js/unit_tests.ts]: Very nice.

File: src/fs.rs [hunk 4]
```diff
@@ -1,13 +1,42 @@
+extern crate rand;
```
Comment (ry) [inline src/fs.rs]: extern calls should be in `src/main.rs` (I think?)

File: src/fs.rs [hunk 5]
```diff
@@ -1,13 +1,42 @@
...
+) -> std::io::Result<PathBuf> {
+  let prefix_ = prefix.unwrap_or("");
+  let suffix_ = suffix.unwrap_or("");
+  let mut buf: PathBuf = match dir {
+    Some(ref p) => p.to_path_buf(),
+    None => std::env::temp_dir(),
+  }.join("_");
+  loop {
+    let unique = thread_rng().next_u32();
+    buf.set_file_name(format!("{}{:08x}{}", prefix_, unique, suffix_));
```
Comment (ry) [inline src/fs.rs]: We have this `tempfile` crate already built into Deno... why not use that? https://docs.rs/tempfile/3.0.3/tempfile/struct.TempDir.html

If we don't want to use that, can we eliminate it and only use this function?

File: src/handlers.rs [hunk 6]
```diff
@@ -395,6 +402,42 @@ fn send_timer_ready(d: *const DenoC, timer_id: u32, done: bool) {
...
+  let deno = from_c(d);
+  if !deno.flags.allow_write {
+    let err = std::io::Error::new(
+      std::io::ErrorKind::PermissionDenied,
+      "allow_write is off.",
+    );
+    return Err(err.into());
+  }
+  // TODO(piscisaureus): use byte vector for paths, not a string.
+  // We can't assume that paths are always valid utf8 strings.
```
Comment (ry) [inline src/handlers.rs]: Could you create an issue for this and link to it here? This is a bigger issue than just makeTempDir.

File: src/fs.rs (hunk 5 above)
Comment (piscisaureus) [reply src/fs.rs]: Tempfile deletes the directory when it goes out of scope.
I looked into it but it's of no use.

File: js/unit_tests.ts (hunk 1 above)
Comment (sijad) [reply js/unit_tests.ts]: shouldn't be?

```ts
assert(!!err);
assertEqual(err.name, "deno.PermissionDenied");
```

Comment (piscisaureus) [reply js/unit_tests.ts]: done (both suggestions)

File: js/unit_tests.ts (hunk 2 above)
Comment (piscisaureus) [reply js/unit_tests.ts]: ok. added

File: js/unit_tests.ts (hunk 3 above)
Comment (piscisaureus) [reply js/unit_tests.ts]: Thanks

File: src/fs.rs (hunk 4 above)
Comment (piscisaureus) [reply src/fs.rs]: fixed

File: src/handlers.rs (hunk 6 above)
Comment (piscisaureus) [reply src/handlers.rs]: done

**Pull request merged**